File: go-version/leetcode/LC2680/maximum_or.go
package LC2680

/**
 * @Author: cutejiuge cutejiuge@163.com
 * @Date: 2025/3/21 下午10:34
 * @FilePath: maximum_or
 * @Description: 位运算实现技巧
 */

func maximumOr(nums []int, k int) int64 {
	var allOr, fixed int // allOr表示所有数字按位或结果，fixed表示固定的1按位组成的结果
	for _, num := range nums {
		fixed |= allOr & num
		allOr |= num
	}

	var ans int64
	for _, num := range nums {
		ans = max(ans, int64((allOr^num)|fixed|num<<k))
	}
	return ans
}

File: python-version/LC2680/maximum_or.py
class Solution:
    def maximumOr(self, nums: list[int], k: int) -> int:
        all_or, fixed = 0, 0
        for num in nums:
            fixed |= all_or & num  # 将固定位的1存放在fixed
            all_or |= num  # 将所有数字的or存放在all_or

        ans = 0
        for num in nums:
            ans = max(ans, (all_or ^ num) | fixed | (num << k))
            # 在all_or中去掉num的或，保留固定位的1,再将num左移k次或上去
        return ans
